stdlib/public/core/Integers.swift.gyb

```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
+  public static var min: DoubleWidth {
     return self.init((High.min, Low.min))
   }
 
-  public static var bitWidth : Int { return 2 * T.bitWidth }
+  public static var bitWidth: Int {
+    return 2 * Base.bitWidth
+  }
 
 % # This covers + and -
 % for x in binaryArithmetic['Numeric'][:2]:
-%   highAffectedByLowOverflow = 'T.max' if x.operator == '+' else 'T.min'
-  public func ${x.name}ReportingOverflow(_ rhs: DoubleWidth<T>)
-    -> (partialValue: DoubleWidth<T>, overflow: ArithmeticOverflow) {
+%   highAffectedByLowOverflow = 'Base.max' if x.operator == '+' else 'Base.min'
+  public func ${x.name}ReportingOverflow(_ rhs: DoubleWidth)
+    -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
     let (low, lowOverflow) =
       _storage.low.${x.name}ReportingOverflow(rhs._storage.low)
     let (high, highOverflow) =
       _storage.high.${x.name}ReportingOverflow(rhs._storage.high)
     let isLowOverflow = lowOverflow == .overflow
-    let result = (high ${x.operator} (isLowOverflow ? 1 : 0), low)
+    let result = (high &${x.operator} (isLowOverflow ? High(1) : High(0)), low)
     let overflow = ArithmeticOverflow(
       highOverflow == .overflow ||
       high == ${highAffectedByLowOverflow} && isLowOverflow
     )
-    return (partialValue: DoubleWidth<T>(result),
+    return (partialValue: DoubleWidth(result),
       overflow: overflow)
   }
 % end
 
-
   public func multipliedReportingOverflow(
-    by rhs: DoubleWidth<T>
-  ) -> (partialValue: DoubleWidth<T>, overflow: ArithmeticOverflow) {
-    let isNegative = (self < DoubleWidth<T>()) != (rhs < DoubleWidth<T>())
-
-    func mul(_ x: Low, _ y: Low, _ carry: Low) -> (partial: Low, carry: Low) {
-      let pair = x.multipliedFullWidth(by: y)
-      let t = DoubleWidth<Low>(pair) + DoubleWidth<Low>((0, carry))
-      return (partial: t._storage.low, carry: t._storage.high)
+    by rhs: DoubleWidth
+  ) -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
+    let (carry, product) = multipliedFullWidth(by: rhs)
+    let result = DoubleWidth(extendingOrTruncating: product)
+    
+    let isNegative = (self < 0) != (rhs < 0)
+    let overflow = isNegative
+      ? carry != ~0
+      : carry != 0
+    return (result, ArithmeticOverflow(overflow))
+  }
+
+  public func quotientAndRemainder(dividingBy other: DoubleWidth)
+    -> (quotient: DoubleWidth, remainder: DoubleWidth) {
+    let isNegative = (self < 0) != (other < 0)
+
+    let rhs = other.magnitude
+    var q = self.magnitude
+
+    // Bail if |other| > |self|
+    if rhs.leadingZeroBitCount < q.leadingZeroBitCount {
+      return (0, self)
     }
+    
+    // Calculate the number of bits before q and rhs line up,
+    // we can skip that many bits of iteration.
+    let initialOffset = q.leadingZeroBitCount + (DoubleWidth.bitWidth - rhs.leadingZeroBitCount) - 1
+
+    // Start with remainder capturing the high bits of q.
+    var r = q >> Magnitude(DoubleWidth.bitWidth - initialOffset)
```

Comment:
It would be interesting to have some benchmarks and the effect of using `>>` vs `&>>` here.